engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/hydration/NadelHydrationInputBuilder.kt

```diff
@@ -1,172 +1,159 @@
-import graphql.nadel.enginekt.util.mapFrom
+import graphql.nadel.enginekt.util.makeNormalizedInputValue
+import graphql.nadel.enginekt.util.toMapStrictly
+import graphql.nadel.enginekt.util.javaValueToAstValue
 import graphql.normalized.ExecutableNormalizedField
 import graphql.normalized.NormalizedInputValue
-import graphql.schema.GraphQLArgument
-import graphql.schema.GraphQLTypeUtil
 
-internal typealias AnyAstValue = Value<*>
+internal class NadelHydrationInputBuilder private constructor(
+    private val instruction: NadelHydrationFieldInstruction,
+    private val aliasHelper: NadelAliasHelper,
+    private val fieldToHydrate: ExecutableNormalizedField,
+    private val parentNode: JsonNode,
+) {
+    companion object {
+        fun getInputValues(
+            instruction: NadelHydrationFieldInstruction,
+            aliasHelper: NadelAliasHelper,
+            fieldToHydrate: ExecutableNormalizedField,
+            parentNode: JsonNode,
+        ): List<Map<String, NormalizedInputValue>> {
+            return NadelHydrationInputBuilder(instruction, aliasHelper, fieldToHydrate, parentNode)
+                .build()
+        }
+    }
 
-internal object NadelHydrationInputBuilder {
-    fun getInputValues(
-        instruction: NadelHydrationFieldInstruction,
-        aliasHelper: NadelAliasHelper,
-        hydratedField: ExecutableNormalizedField,
-        parentNode: JsonNode,
-    ): List<Map<String, NormalizedInputValue>> {
-        val inputDefsForAllCalls = instruction.actorInputValueDefs.asSequence()
-            .let { defs ->
-                when (val hydrationStrategy = instruction.hydrationStrategy) {
-                    is NadelHydrationStrategy.OneToOne -> defs
-                    is NadelHydrationStrategy.ManyToOne -> defs.filter { def -> def != hydrationStrategy.inputDefToSplit }
-                }
-            }
+    private fun build(): List<Map<String, NormalizedInputValue>> {
+        return when (val hydrationStrategy = instruction.hydrationStrategy) {
+            is NadelHydrationStrategy.OneToOne -> makeOneToOneArgs()
+            is NadelHydrationStrategy.ManyToOne -> makeManyToOneArgs(hydrationStrategy)
+        }
+    }
 
-        val argsForAllCalls = mapFrom(
-            inputDefsForAllCalls
-                .mapNotNull { actorInputDef ->
-                    val argumentDef = instruction.actorFieldDef.getArgument(actorInputDef.name)
-                    val inputValue = makeInputValue(
-                        actorInputDef = actorInputDef,
-                        argumentDef = argumentDef,
-                        parentNode = parentNode,
-                        hydrationField = hydratedField,
-                        aliasHelper = aliasHelper,
-                    ) ?: return@mapNotNull null
-                    actorInputDef.name to inputValue
-                }
-                .toList()
+    private fun makeOneToOneArgs(): List<Map<String, NormalizedInputValue>> {
```

Comment:
Ok so if you're not aware, we support two types of plain hydration.

The usual

```graphql
type Query {
  userById(id: ID): User
}
type Issue {
  userId: ID
  user: User @hydrated(from: "userById" arguments: [
    {name: "id" valueFromField: "userId"}
  ])
}
```

i.e. one to one. But then we also have


```graphql
type Query {
  userById(id: ID): User
}
type Issue {
  userIds: [ID] # NOTE the array here
  user: User @hydrated(from: "userById" arguments: [
    {name: "id" valueFromField: "userIds"}
  ])
}
```

That is, for N `userIds` we need to send N queries to the underlying service i.e. many to one.

Not a great pattern because you should really be using batching, but we support it.

Reply:
Ah ok, this answers the question I've asked bellow (I usually review the tests first).